### ISSUE TITLE
Using 'O' (shit-o) to insert line above is not taking the current line's indentation into account

### ISSUE THREAD

**Issue by rlofc** [mawww/kakoune#894]
This seems inconsistent with the behavior of using 'o' to insert line below. Given the following code:
```
int main() { 
    int foo; 

    int bar; 
}
```
Place your cursor on the 'i' of int in line 4 and press 'O' (shift-o). A line will be added and the cursor will be positioned at the beginning of the row. 
This feels wrong, as I believe it should take the current line's indentation into account and align with its indentation at column 4.
When using 'o' to insert line below, the behavior is correct and the cursor is positioned taking the current line's indentation into account.

**Comment by mawww** [comment]
Hello, this is not specific to 'O', but just a result of the indentation logic (in `c-family.kak`) that copies the indentation from the previous line (which in yourcase is empty). If you position your selection on line 3, and i<ret>, you'll get the same result.

We could tweak the indentation logic so that it copies the indentation of the previous non empty line to fix that.

**Comment by rlofc** [comment]
Yes, that would work better, but not for all cases, for example:
```
void foo() {

    int bar;
}
```
When I'm positioned on the 'b' and I press <shift-o>, I would prefer to have the same indentation as my current line.
However, this is not the case when I'm positioned on the '}', in which case, getting the indentation from the last non-empty line works better.
I guess to make this cover all cases the new line indentation logic has to be aware of block boundries as well.

**Comment by lenormf** [comment]
The goal of the indent hooks isn't to provide a full proof indentation logic, they are simple helpers that just use a minimal low-range context and make a decision based on it. Anything that requires parsing code (and detecting blocks for instance) goes beyond their purpose, it would be too much work to implement/maintain the hooks if they did. Use a formatter (e.g. `clang-format`) if you want advanced formatting capabilities.

**Comment by mawww** [comment]
> When I'm positioned on the 'b' and I press `<shift-o>`, I would prefer to have the same indentation as my current line.

We can have that as well, it wont be based on the line you were in though, but on the fact that we are just after an opening `{`. The same thing is happening here, we are only looking at the immediate previous line for an openging `{`, we could be looking at the previous non-empty line.

**Comment by rlofc** [comment]
Yes, sounds good. If this is a reasonable fix to the c-family.kak then I think it's worth doing and could potentially help vim-spoiled newcommers such as myself.

**Comment by lenormf** [comment]
I think this issue can be closed, the hook handles that now AFAIK.
